Fix StoppableTask to allow for proper select() based event loops, fix MQTT

dwmw2 opened the pull request:
Worker threads need to wake up quickly and respond to three event types:
 - When work comes in on a socket or serial port
 - When they're asked to stop
 - When a certain time is reached (for periodic events, heartbeats, etc.)

Domoticz currently doesn't provide a way for threads to act responsively on all three. Mostly we use `StoppableTask::IsStopRequested()` which will wait for a period of time and return immediately if the thread is asked to stop _(which isn't the common case we should be optimising for)_. And then we process the actual work of the thread mostly as an afterthought, often up to a second after it actually arrived.

The alternative approach, like in `MainWorker::Do_Work_On_Rx_Messages()`, is to prioritise the *work* instead. It mostly spends its time sitting in `m_rxMessageQueue.timed_wait_and_pop()`, waking every five seconds just to check if it should stop with an `IsStopRequested(0)` call.

The current `StoppableTask` implementation means that either the work, or the stop check, is done as an 'afterthought' and not handled promptly. And it means that threads are waking up often when they don't need to, as well as not waking up as fast as they should when they *do*.

This change to `StoppableTask` makes it possible for a thread to react quickly to all three. It changes the implementation to use a socketpair which means that we can add a `GetStopFd()` method, and then a worker thread can use a standard `select()` loop waiting on that fd as well as its file descriptors for real I/O, with whatever timeout it desires.

Then we can fix MQTT to use a proper select-based loop, so #6433 is fixed and MQTT updates are seen fairly much instantly without needing the hack of an extra thread (although I've left that fix out of this PR).

The EventSystem and Scheduler are OK with the existing `IsStopRequested()` interface because they don't have separate file descriptors to poll on, but they were waking up far more often than they needed to. EventSystem was waking twice a second just to see if it had reached the time that it needed to do something. So we give it a hint: if it doesn't have anything to do for 12 seconds, then it can sleep for 12 seconds in `IsStopRequested()`. Aside from fewer wakeups, this actually improves the precision of when the actions happen anyway; they're now within a millisecond of the top of the minute, instead of up to half a second thereafter.

Scheduler was similar although for now I've left it waking up every second because although most scheduler entries only have minute precision, I can't prove to myself that there's no way to schedule something at a precise second _(e.g. sunset/sunrise based?)_ But the cleanup does improve the precision still, so events are run at the start of the second, not at some random point within each second.

dwmw2 committed:
Add socketpair support to StoppableTask for efficient event-driven I/O

Worker threads need to wake up quickly and respond to three event types:
 - When work comes in on a socket or serial port
 - When they're asked to stop
 - When a certain time is reached (for periodic events, heartbeats, etc.)

Domoticz currently doesn't provide a way for threads to act responsively
on all three. Mostly we use StoppableTask::IsStopRequested() which will
wait for a period of time and return immediately if the thread is asked
to stop (which isn't the common case we should be optimising for). And
then we process the actual work of the thread mostly as an afterthought,
often up to a second after it actually arrived.

The alternative approach, like in MainWorker::Do_Work_On_Rx_Messages(),
is to prioritise the *work* instead. It mostly spends its time sitting in
m_rxMessageQueue.timed_wait_and_pop(), waking every five seconds just to
check if it should stop with an IsStopRequested(0) call.

The current StoppableTask implementation means that either the work, or
the stop check, is done as an 'afterthought' and not handled promptly.
And it means that threads are waking up often when they don't need to,
as well as not waking up as fast as they should when they *do*.

This change to StoppableTask makes it possible for a thread to react
quickly to all three. It changes the implementation to use a socketpair
which means that we can add a GetStopFd() method, and then a worker
thread can use a standard select() loop waiting on that fd as well as
its file descriptors for real I/O, with whatever timeout it desires.

gizmocuz commented:
Thank you for this PR... Could you undo the changes to MQTT as they conflict with my changes in #6430 ...

dwmw2 commented:
I suspect the MQTT thread changes probably didn't live in a CounterHelper PR? I also don't actually think the extra queue is necessary at all if we fix the worker thread to actually wake up when it has work to do :)

I'll keep it here for my testing for now, and I can rebase on top if you do merge the other changes, OK?

dwmw2 commented:
Moving some comments from Slack so they don't get lost..

> Would it be a good idea to move (later) everything to the header, or move all code to the .cpp? (like constructor)

Happy either way. I moved it all out to a .cpp file so I could play with the implementation without having to rebuild every C++ file in the project, but once I'm done playing with it that's no longer needed. 

> You do not need all those if/defs, unix should/could also use closesocket/send/recv
> Also please include stdafx.h like in all .cpp files, then there should be no need for all these includes as most of them are already in that header file

Not closesocket unless we have a wrapper? I'll take a look. Mostly I threw that in as known good code and went to concentrate on fixing MQTT, so I'll look at making it more "Domoticzy". And I'll use the CI here to check the Windows build.

dwmw2 committed:
Reduce unnecessary wakeups in EventSystem

Calculate sleep time until next 12-second heartbeat instead of waking
every 500ms. Since 60 is a multiple of 12, this also wakes at the top
of each minute. Reduces wake-ups from ~120/minute to ~5/minute and
improves timing precision for minute-boundary events from ~500ms jitter
to millisecond precision.

dwmw2 committed:
Reduce unnecessary wakeups in Scheduler

Calculate sleep time until next event (next second or 12-second heartbeat)
instead of waking every second unconditionally. Still wakes ~60 times/minute
for schedule checking (needed for randomness offsets), but timing is more
precise and code is cleaner.

CheckSchedules() runs every second. DeleteExpiredTimers() runs only at
minute boundaries.

dwmw2 commented:
Reworked and cleaned up the Windows code in StoppableTask.cpp. Now it *only* uses szUserDataFolder as the directory for its AF_UNIX sockets. That's going to be writeable. It doesn't leave anything behind because it deletes the listening socket as soon as it's created the connected socketpair that it needs. And it keeps the fallback to AF_INET so *if* you run Domoticz so it can't write to use szUserDataFolder... well, it probably won't get far but at least *this* code will cope :)

Test case which shows the difference in speed of worker threads actually waking up for *work*: https://david.woodhou.se/MagWeg2-work.zip

Compile with and without `-DUSE_GETSTOPFD` to see the difference.
```
 ● I will run the following shell command: 
 ./MagWeg2
 ⋮ 
 ↳ Purpose: Run test with USE_GETSTOPFD

Sending ping 1...
Entered thread function
Pong received, latency: 126 µs
Sending ping 2...
Pong received, latency: 132 µs
Sending ping 3...
Pong received, latency: 76 µs
Sending ping 4...
Pong received, latency: 173 µs
Sending ping 5...
Pong received, latency: 152 µs
main app stopping thread....
Stop requested
Leaving thread function
main app thread stopped....
 ⋮ 
 ● Completed in 3.513s





🛠️  Using tool: execute_bash (trusted)
 ⋮ 
 ● I will run the following shell command: 
 ./MagWeg2-poll
 ⋮ 
 ↳ Purpose: Run test without USE_GETSTOPFD

Sending ping 1...
Entered thread function
Pong received, latency: 1001233 µs
Sending ping 2...
Pong received, latency: 299991 µs
Sending ping 3...
Pong received, latency: 300785 µs
Sending ping 4...
Pong received, latency: 301022 µs
Sending ping 5...
Pong received, latency: 300776 µs
main app stopping thread....
Leaving thread function
main app thread stopped....
 ⋮ 
 ● Completed in 5.719s


> ## Test Results Summary

With USE_GETSTOPFD (event-driven with select()):
• Ping 1: 126 µs
• Ping 2: 132 µs
• Ping 3: 76 µs
• Ping 4: 173 µs
• Ping 5: 152 µs
• Average: ~132 µs (0.132 milliseconds)

Without USE_GETSTOPFD (polling with 1-second IsStopRequested()):
• Ping 1: 1,001,233 µs (1.001 seconds)
• Ping 2: 299,991 µs (0.300 seconds)
• Ping 3: 300,785 µs (0.301 seconds)
• Ping 4: 301,022 µs (0.301 seconds)
• Ping 5: 300,776 µs (0.301 seconds)
• Average: ~441 milliseconds

Performance improvement: ~3,300x faster with the event-driven approach using GetStopFd(). The polling version must wait up to 1 second in IsStopRequested() before checking for
work, while the event-driven version responds immediately via select().

```

gizmocuz commented:
Thank you for this PR, time and patience 😁
